### ISSUE TITLE
Expand use of __ro_after_init

### ISSUE THREAD

**Issue by kees** [KSPP/linux#15]
There continues to be a lot of things in `__init` or `__initdata` areas that are only ever written to once, and never change again. It would be nice to find a way to automatically discover these cases and mark them with `__ro_after_init` to reduce the attack surface within the kernel.

**Comment by ihalip** [comment]
I added a very crude `clang-tidy` check in my local llvm clone for this specific case: __initdata variables that are only assigned to in __init functions. This found 275 instances of variables that are possible candidates to mark as __ro_after_init (with a fraction of that number being false positives): https://gist.github.com/ihalip/1eb493e4805aca9be06999885ad7d3b6

**Comment by kees** [comment]
Oh very nice! Can you generate a huge patch and we could split it up with https://github.com/kees/kernel-tools/blob/master/split-on-maintainer

**Comment by ihalip** [comment]
Yeah - it'll take a while though, I'm replacing the macros manually because I didn't find a simple way to do with `clang-tidy`.

**Comment by ihalip** [comment]
Oh, I think I was very wrong - what I found with my check was `__initdata` variables that could be marked `__ro_after_init`. But doing so moves them into `.rodata` and they don't get discarded by `free_initmem()`. Let me take another look...

**Comment by kees** [comment]
Yeah, __initdata variables go away after __init, so they're not useful for this. You want variables that are _not_ __initdata and are only written to from __init functions (or only during a __init lifetime -- written in a non-init function but only when called from an __init function).

**Comment by kees** [comment]
Here are various __ro_after_init candidates identified by GrapheneOS:

https://github.com/GrapheneOS/kernel_google_coral/commit/7da16e3692bb9f9beb80993e87e13bccc90ed957
https://github.com/GrapheneOS/kernel_google_coral/commit/72f72b7407f813ae7a163f9e3a2a2403f04e3084
https://github.com/GrapheneOS/kernel_google_coral/commit/cc822d47879f4c2a470fb42bf0bd66a93bc5685f
https://github.com/GrapheneOS/kernel_google_coral/commit/3b392dcaf071e54ffd69be2b6fd46ac535406a25
https://github.com/GrapheneOS/kernel_google_coral/commit/68c5d9a7774da5ac1a325f93da39d06757c63a2d
https://github.com/GrapheneOS/kernel_google_coral/commit/7ed06f7fc5845e70d914d7ad1f0ec72cf635773f
https://github.com/GrapheneOS/kernel_google_coral/commit/2f1f03b258d5e2255a3a8ce33df459ad79a7ee3c
https://github.com/GrapheneOS/kernel_google_coral/commit/c67bc988bb079e3724d183bcd4097bd4f840ee00

**Comment by icb-linux** [comment]
I've been working on a small Clang-based tool to find possible `__ro_after_init` variables. It reads the kernel’s compile_commands.json and looks for global variables that appear to be written only during initialization. It follows calls into helper functions and notes cases that need extra care, such as pointers, callbacks, assembly, and cross-file calls. Each finding has a high, medium, or low confidence level, along with source evidence and an explanation. The tool is meant to guide human review, not prove that every candidate is safe.

The tool uses Clang to parse each configured C file and inspect its syntax tree. It records global variables, writes to those variables, and the function calls that can reach each write. It then checks whether every observed write happens on an initialization-only call path. Cases involving indirect writes, external symbols, callbacks, assembly, or incomplete call information receive a lower confidence score.

The attached report contains the results of a full scan of all 2,887 translation units enabled by the Linux 7.1.5 x86_64_defconfig configuration at commit 155b42bec9cbb6b8cdc47dd9bd09503a81fbe493.

[linux-7.1.5-full-x86_64-defconfig.json](https://github.com/user-attachments/files/30358314/linux-7.1.5-full-x86_64-defconfig.json)

I'm still working on some issues with the code, most of which was written by an LLM coding agent. But I figured I'd include the kind of output I'm generating to see if there is any interest in using it to guide some of this work.
